mp/src/game/client/momentum/ui/HUD/hud_speedometer_label.cpp

```diff
@@ -0,0 +1,248 @@
+{
+    CreateComparisonLabel();
+    Reset();
+}
+
+// creates a separate label for the comparison & pins it to this
+void SpeedometerLabel::CreateComparisonLabel()
+{
+    char name[BUFSIZELOCL];
+    Q_snprintf(name, sizeof(name), "%sComparison", GetName());
+    m_pComparisonLabel = new Label(GetParent(), name, "");
+    m_pComparisonLabel->SetName(name);
+    m_pComparisonLabel->SetPos(0, 0);
+    m_pComparisonLabel->PinToSibling(GetName(), PIN_TOPLEFT, PIN_TOPRIGHT);
+    m_pComparisonLabel->SetFont(GetFont());
+    m_pComparisonLabel->SetAutoTall(true);
+    m_pComparisonLabel->SetAutoWide(true);
+    m_pComparisonLabel->SetContentAlignment(a_center);
+    m_pComparisonLabel->SetPinCorner(PIN_TOPLEFT, 0, 0);
+}
+
+void SpeedometerLabel::ApplySchemeSettings(IScheme *pScheme)
+{
+    BaseClass::ApplySchemeSettings(pScheme);
+    m_NormalColor = GetSchemeColor("MOM.Speedometer.Normal", pScheme);
+    m_IncreaseColor = GetSchemeColor("MOM.Speedometer.Increase", pScheme);
+    m_DecreaseColor = GetSchemeColor("MOM.Speedometer.Decrease", pScheme);
+    m_pComparisonLabel->SetFont(GetFont());
+}
+
+void SpeedometerLabel::PerformLayout()
+{
+    BaseClass::PerformLayout();
+
+    // keep centered in parent
+    char szMain[BUFSIZELOCL];
+    GetText(szMain, BUFSIZELOCL);
+
+    HFont labelFont = GetFont();
+    int combinedLength = UTIL_ComputeStringWidth(labelFont, szMain);
+
+    if (m_eColorizeType == SPEEDOMETER_COLORIZE_COMPARISON_SEPARATE && m_bDrawComparison)
+    {
+        char szComparison[BUFSIZELOCL];
+        m_pComparisonLabel->GetText(szComparison, BUFSIZELOCL);
+        combinedLength += UTIL_ComputeStringWidth(labelFont, szComparison);
+    }
+
+    int xOffset = (GetParent()->GetWide() - combinedLength) / 2;
+    SetPos(xOffset, GetYPos());
+}
+
+void SpeedometerLabel::OnThink()
+{
+    if (HasFadeOutAnimation() && !m_bDoneFading)
+    {
+        SetAlpha(*m_pFlAlpha);
+        m_pComparisonLabel->SetAlpha(*m_pFlAlpha);
+
+        if (CloseEnough(*m_pFlAlpha, 0.0f, FLT_EPSILON))
+        {
+            m_bDoneFading = true;
+            Reset();
+        }
+    }
+}
+
+void SpeedometerLabel::Reset()
+{
+    m_flCurrentValue = 0.0f;
+    m_flPastValue = 0.0f;
+    m_flDiff = 0.0f;
+    m_flNextUpdateCheck = 0.0f;
+
+    m_bCustomDiff = false;
+    m_bDrawComparison = true;
+    m_bDoneFading = true;
+
+    m_pComparisonLabel->SetText("");
+    BaseClass::SetText("");
+
+    SetAlpha(HasFadeOutAnimation() ? 0 : 255);
+}
+
+void SpeedometerLabel::SetText(int value)
+{
+    char szValue[BUFSIZELOCL];
+    Q_snprintf(szValue, sizeof(szValue), "%i", value);
+    BaseClass::SetText(szValue);
+}
+
+void SpeedometerLabel::Update(float value)
+{
+    if (m_flNextUpdateCheck > gpGlobals->curtime)
+        return;
+
+    m_flCurrentValue = value;
+
+    ConvertUnits();
+    SetText(m_flCurrentValue);
+    Colorize();
+    m_bDoneFading = HasFadeOutAnimation() ? !StartFadeout() : false;
```

Comment:
Why is this done here as well? Isn't this updated in the `OnThink` method above?

Reply:
This stops and starts the fadeout when the label is updated, while `OnThink` handles constantly setting the alpha of it & resetting it's values when it's done fading out.